content/developer/embeds/index.md

```diff
@@ -0,0 +1,100 @@
+* `canHandle` - Given both a domain and a full URL, this method determines if a URL can be handled by the embed class. If the result indicates it cannot handle a URL, subsequent calls to this embed for this URL are halted. If the result indicates the embed class can handle the URL, it is assumed a call to `matchUrl` is safe and will return relevant data.
+* `matchUrl` - This method is expected to obtain relevant information about a URL. This does not necessarily mean the information has to be found in the document the URL points to. For example, the YouTube embed class retrieves the video ID from the URL, then fetches a separate oEmbed endpoint to get specific information about that video. Data returned from this endpoint should either be an associative array or `null`, if unable to obtain relevant information. If an array, the following keys are allowed:
+    * name: Name of the resource (e.g. video's title).
+    * body: Body text, if any (e.g. video description).
+    * photoUrl: Associated image URL (e.g. video thumbnail).
+    * height: Height of the resource (e.g. video height).
+    * width: Width of the resource (e.g. video width).
+    * attributes: An array of arbitrary associated data (e.g. a video ID).
+* `renderData` - Embed markup generated for inclusion on a page is expected to be returned by this method. The data array parameter should be the same as was returned by the `matchUrl` method in the class.
+
+### Registering the Embed Component
+
+Once an embed class has been created, it must be registered with the embed manager. The embed manager is the centralized handler of what embed classes are called, based on a URL. New embeds must be added with the `EmbedManager::addEmbed` method, called on the shared instance. This requires hooking early into Vanilla's setup, but after the dependency injection container has been initialized. The `container_init` event can be used to perform this action.
+
+```php
+public function container_init_handler(Container $dic) {
+    $dic->rule(Vanilla\Embeds\EmbedManager::class)
+        ->addCall('addEmbed', [new Reference(MyEmbed::class)])
+}
+```
+
+After the embed class has been registered, subsequent calls to the embed manager will use the class for URL scraping and markup generation. To test, you can use Vanilla's API v2 media endpoint to scrape a URL that should be handled by the new component.
+
+## Client-Side Components
+
+Client-side rendering is performed with JavaScript. The implementation of a client-side embed renderer is straightforward, primarily powered by a single renderer function.
+
+Vanilla's `registerEmbed` function is required to register the new renderer. It must be imported from `@dashboard/embeds`. In addition, the `IEmbedData` interface is also required for  constructing a proper render function signature. To import these two requirements, you may include the following:
+
+```javascript
```

Comment:
These should all be changed to `typescript` or `ts`. The type definitions in the function signatures and interfaces are not part of javascript. I think we need to make that pretty